File: projects/Vibe_XASR/windows_build/src/VibeXASR.Windows/Lexicon/PinyinNormalizer.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace VibeXASR.Windows.Lexicon;

/// <summary>
/// Homophone correction by pinyin: rewrite a run of CJK chars that SOUNDS like a dictionary word
/// into that word's exact spelling (e.g. 贾阳清/嘉阳清/贾杨青 → 贾扬清). Faithful port of the macOS
/// PinyinNormalizer. Loads a char→toneless-pinyin table (pinyin.txt); only the user's multi-char,
/// all-CJK dictionary words drive it (longest-first, fuzzy tone-insensitive matching). Inert when
/// the table or word list is empty.
/// </summary>
public sealed class PinyinNormalizer
{
    private Dictionary<char, HashSet<string>> _table = new();
    private bool _loaded;
    // Dictionary words (CJK only), longest first, with each char's fuzzy-reading set precomputed.
    private List<(char[] Word, HashSet<string>[] Reads)> _words = new();
    private const int MinLen = 2;

    public bool IsActive => _loaded && _words.Count > 0;

    /// <summary>Load the 汉字→拼音 table once (format: "char py1 py2 …"). No-op if missing.</summary>
    public void LoadTableIfNeeded(string? path)
    {
        if (_loaded || string.IsNullOrEmpty(path) || !File.Exists(path)) return;
        try
        {
            var t = new Dictionary<char, HashSet<string>>(28000);
            foreach (var line in File.ReadLines(path))
            {
                var parts = line.Split(' ', StringSplitOptions.RemoveEmptyEntries);
                if (parts.Length == 0 || parts[0].Length == 0) continue;
                t[parts[0][0]] = new HashSet<string>(parts.Skip(1));
            }
            _table = t;
            _loaded = true;
        }
        catch (Exception ex) { Diag.Log("pinyin table load failed: " + ex.Message); }
    }

    private static bool IsCjk(char c) =>
        (c >= 0x3400 && c <= 0x4DBF) || (c >= 0x4E00 && c <= 0x9FFF) || (c >= 0xF900 && c <= 0xFAFF);

    /// <summary>Collapse the common accent confusions so they compare equal:
    /// zh/ch/sh→z/c/s, n…→l…, ing/eng/ang→in/en/an.</summary>
    public static string FuzzyKey(string p)
    {
        var s = p;
        if (s.StartsWith("zh") || s.StartsWith("ch") || s.StartsWith("sh")) s = s[0] + s.Substring(2);
        if (s.StartsWith("n")) s = "l" + s.Substring(1);
        if (s.EndsWith("ing")) s = s.Substring(0, s.Length - 3) + "in";
        else if (s.EndsWith("eng")) s = s.Substring(0, s.Length - 3) + "en";
        else if (s.EndsWith("ang")) s = s.Substring(0, s.Length - 3) + "an";
        return s;
    }

    private HashSet<string> FuzzySet(char c) =>
        _table.TryGetValue(c, out var r) ? new HashSet<string>(r.Select(FuzzyKey)) : new HashSet<string>();

    /// <summary>Set the active dictionary words — only multi-char, all-CJK words are used.</summary>
    public void SetWords(IEnumerable<string> raw)
    {
        _words = raw
            .Select(w => w.ToCharArray())
            .Where(chars => chars.Length >= MinLen && chars.All(IsCjk))
            .Select(chars => (Word: chars, Reads: chars.Select(FuzzySet).ToArray()))
            .OrderByDescending(t => t.Word.Length)
            .ToList();
    }

    /// <summary>Rewrite homophone runs into their dictionary spelling.</summary>
    public string Normalize(string text)
    {
        if (!IsActive) return text;
        var chars = text.ToCharArray();
        var outp = new List<char>(chars.Length);
        int i = 0;
        while (i < chars.Length)
        {
            bool matched = false;
            foreach (var (word, reads) in _words) // longest first
            {
                int L = word.Length;
                if (i + L > chars.Length) continue;
                if (MatchesExact(chars, i, word)) continue; // already exact → leave it
                bool ok = true;
                for (int k = 0; k < L; k++)
                {
                    var r = FuzzySet(chars[i + k]);
                    if (r.Count > 0 && reads[k].Count > 0 && r.Overlaps(reads[k])) continue;
                    ok = false; break;
                }
                if (ok) { outp.AddRange(word); i += L; matched = true; break; }
            }
            if (!matched) { outp.Add(chars[i]); i++; }
        }
        return new string(outp.ToArray());
    }

    private static bool MatchesExact(char[] chars, int i, char[] word)
    {
        for (int k = 0; k < word.Length; k++) if (chars[i + k] != word[k]) return false;
        return true;
    }
}

File: projects/Vibe_XASR/macos_build/native/app/Sources/VibeIME/PinyinNormalizer.swift
import Foundation

/// Homophone correction by pinyin: rewrite any run of CJK chars that *sounds the
/// same* as a dictionary word into that word's exact spelling. This catches what
/// contextual biasing can't — a proper noun like "贾扬清" that the recognizer
/// scatters across same-sounding chars (贾阳清 / 嘉阳清 / 贾杨青 / 贾央青 …). All
/// share the toneless pinyin "jia yang qing", so they normalize to the one
/// spelling the user put in their dictionary.
///
/// Matching is per-character, multi-reading (uses the char's full pinyin set so
/// heteronyms still match), longest dictionary word first. Only the user's CJK
/// dictionary words drive it, so the scope is exactly what they opted into.
@MainActor
final class PinyinNormalizer {
    static let shared = PinyinNormalizer()

    /// char → set of toneless readings (e.g. 沈 → {shen, chen, tan}).
    private var table: [Character: Set<String>] = [:]
    private var loaded = false

    /// Dictionary words (CJK only), longest first, with each word's per-char
    /// reading sets precomputed.
    private var words: [(word: [Character], reads: [Set<String>])] = []

    private init() {}

    /// Load the bundled 汉字→拼音 table once. No-op (and normalizer stays inert)
    /// when the table is missing.
    func loadTableIfNeeded(path: String?) {
        guard !loaded, let path, let text = try? String(contentsOfFile: path, encoding: .utf8) else { return }
        var t: [Character: Set<String>] = [:]
        t.reserveCapacity(28000)
        for line in text.split(whereSeparator: \.isNewline) {
            let parts = line.split(separator: " ")
            guard let first = parts.first, let ch = first.first else { continue }
            t[ch] = Set(parts.dropFirst().map(String.init))
        }
        table = t
        loaded = true
    }

    /// Minimum length for a word to drive pinyin normalization. Single chars (and
    /// the empty case) are NEVER allowed — a 1-char homophone rule would rewrite
    /// every same-sounding char in any text (e.g. "沐" would eat 木/目/牧…). 2 is the
    /// floor (2-char names are common); longer is safer still.
    private let minLen = 2

    /// Set the active dictionary words. Only multi-char, all-CJK words are used —
    /// English words and single chars are skipped. Empty list disables it.
    func setWords(_ raw: [String]) {
        words = raw
            .map { Array($0) }
            .filter { chars in chars.count >= minLen && chars.allSatisfy { isCJK($0) } }
            .map { chars in (word: chars, reads: chars.map { self.fuzzySet($0) }) }
            .sorted { $0.word.count > $1.word.count }
    }

    var isActive: Bool { loaded && !words.isEmpty }

    private func isCJK(_ c: Character) -> Bool {
        guard c.unicodeScalars.count == 1, let v = c.unicodeScalars.first?.value else { return false }
        return (0x3400...0x4DBF).contains(v) || (0x4E00...0x9FFF).contains(v) || (0xF900...0xFAFF).contains(v)
    }

    /// Normalize a toneless syllable for FUZZY matching, merging the most common
    /// accent confusions so they compare equal:
    ///   • retroflex/flat: zh/ch/sh → z/c/s   (是 shi ≈ 四 si)
    ///   • l ↔ n:           n… → l…           (牛 niu ≈ 刘 liu)
    ///   • front/back nasal: ing/eng/ang → in/en/an  (京 jing ≈ 金 jin)
    static func fuzzyKey(_ p: String) -> String {
        var s = p
        if s.hasPrefix("zh") || s.hasPrefix("ch") || s.hasPrefix("sh") {
            s.remove(at: s.index(after: s.startIndex))   // drop the 'h'
        }
        if s.hasPrefix("n") { s = "l" + String(s.dropFirst()) }
        if s.hasSuffix("ing") { s = String(s.dropLast(3)) + "in" }
        else if s.hasSuffix("eng") { s = String(s.dropLast(3)) + "en" }
        else if s.hasSuffix("ang") { s = String(s.dropLast(3)) + "an" }
        return s
    }

    /// A char's readings collapsed to fuzzy keys (empty for non-CJK / unknown).
    private func fuzzySet(_ c: Character) -> Set<String> {
        guard let r = table[c] else { return [] }
        return Set(r.map { Self.fuzzyKey($0) })
    }

    /// Rewrite homophone runs into their dictionary spelling.
    func normalize(_ text: String) -> String {
        guard isActive else { return text }
        let chars = Array(text)
        var out: [Character] = []
        out.reserveCapacity(chars.count)
        var i = 0
        while i < chars.count {
            var matched = false
            for (word, reads) in words {     // longest first
                let L = word.count
                guard i + L <= chars.count else { continue }
                if word == Array(chars[i..<i+L]) { continue }   // already exact → leave it
                var ok = true
                for k in 0..<L {
                    // both sides share a FUZZY reading (l/n, zh-z, in-ing… merged)
                    let r = fuzzySet(chars[i + k])
                    if !r.isEmpty, !reads[k].isEmpty, !r.isDisjoint(with: reads[k]) { continue }
                    ok = false; break
                }
                if ok {
                    out.append(contentsOf: word)
                    i += L
                    matched = true
                    break
                }
            }
            if !matched {
                out.append(chars[i])
                i += 1
            }
        }
        return String(out)
    }
}
